blockchain/difficulty.go

```diff
@@ -176,11 +182,31 @@ func (b *BlockChain) calcNextRequiredDifficulty(prevNode *blockNode, newBlockTim
 		nextDiffBig.Set(nextDiffBigMin)
 	}
 
-	// Limit new value to the proof of work limit.
+	// Prevent the difficulty from going lower than the minimum allowed
+	// difficulty.
+	//
+	// Larger numbers result in a lower difficulty, so imposing a minimum
+	// difficulty equates to limiting the maximum target value.
 	if nextDiffBig.Cmp(params.PowLimit) > 0 {
 		nextDiffBig.Set(params.PowLimit)
 	}
 
+	// Prevent the difficulty from going higher than a maximum allowed
+	// difficulty on the test network.  This is to prevent runaway difficulty on
+	// testnet by ASICs and GPUs since it's not reasonable to require
+	// high-powered hardware to keep the test network running smoothly.
+	//
+	// Smaller numbers result in a higher difficulty, so imposing a maximum
+	// difficulty equates to limiting the minimum target value.
+	//
+	// This rule is only active on the version 3 test network once the max diff
+	// activation height has been reached.
+	if b.minTestNetTarget != nil && nextDiffBig.Cmp(b.minTestNetTarget) < 0 &&
+		(!b.isTestNet3() || nextHeight >= testNet3MaxDiffActivationHeight) {
```

Comment:
The `!b.isTestNet3()` check looks redundant here, as `b.minTestNetTarget != nil` will only be true if `b.isTestNet3()` is also true. Is this to somehow futureproof or foolproof a specific case?

Reply:
Yes, I wanted it to continue to work with all future versions of the test network by default.